### DISCUSSION TITLE
Recurring transactions implementation improvement?

### DISCUSSION THREAD

**Discussion by stepanov1975** [firefly-iii/firefly-iii#5702] [General]
Hi.
From what I can see recurring transactions implemented in the following way

1. Recurring transaction fired only if cron call executed in the day it defined in the transaction. If cron missed the day recurring transactions will not happen
2. Protection from multiple executions of the same recurring transaction on the same day implemented by disallowing cron call more than once in 24h

I would propose a different approach. Add to the database for recurring transactions column with the date when the transaction was executed last time. This information is useful anyway. Then each time cron called just check if the transaction was executed at the proper date. If not fire it. This seems to be a more robust approach. Both allow missing cron calls without missing transitions and protect against multiple cron calls on the same day.

**Comment by JC5** [comment]
You can run the cron job as often as you like. It will not create extra transactions.

> Add to the database for recurring transactions column with the date when the transaction was executed last time.

This is already in place.

**Comment by stepanov1975** [comment]
@JC5 
I understand that there is no problem with extra transactions currently. I will try to explain myself once more
I am personally always trying to simplify my code. The simpler code the less chance to have bugs and unexpected behavior.
The current solution to work needs a couple of things to be true
1. Cron called at least once a day
2. If it is called more than once the part of the code that rejects multiple calls works
3. Server time is accurate and has no unexpected changes
4. No crash between the moment you are marking cron as executed and moment transaction created

If you instead will just check if the transaction was created in the day it needs to be created, and create it if not, this is simple solution and doesn't require any other things to happen to work. At least it seems to me so.
Anyway. Excellent program and it helps me a lot to manage my money :) THanks!

**Comment by Nu-hin** [comment]
My server was down for a week. All recurring transactions that were supposed to fire during this week didn't. I agree to @stepanov1975  that either the script should process _all due_ recurring transactions instead of the ones _due today_, or there should at least be a way to manually trigger the skipped ones (currently in the UI there's a possibility to manually fire future transactions, but not the past ones).
